autoAdjustHwloc.hpp

```diff
@@ -23,154 +24,151 @@ class autoAdjust
+	// This will usually be 1 x L3, but can be 2 x L2 per package
+	void proccessTopLevelCache(hwloc_obj_t obj)
+	{
+		if(obj->attr == nullptr)
+			throw(std::runtime_error("Cache object hasn't got attributes."));
 
-				int nativeCores = hwloc_bitmap_weight(cpuset);
-				int numPus = obj->arity;
-				for (int i = 0; i < numPus && numHashes != 0 && firstNativeCore != -1; i++)
-				{
-					hwloc_obj_t pu = obj->children[i];
-					// only use native pu's
-					if (pu->type == HWLOC_OBJ_PU && hwloc_bitmap_isset( cpuset, i + firstNativeCore ))
-					{
-						// if no cache is available we give each native core a hash
-						int numUnit = numCachesLeft != 0 ? numCachesLeft : nativeCores;
-
-						// two hashes per native pu if number of hashes if larger than compute units
-						int power = numHashes > numUnit ? 2 : 1;
-						char strbuf[256];
-	
-						snprintf(strbuf, sizeof(strbuf), "   { \"low_power_mode\" : %s, \"no_prefetch\" : true, \"affine_to_cpu\" : %u },\n",
-							power == 2 ? "true" : "false", pu->os_index);
-						printer::inst()->print_str(strbuf);
-
-						// update number of free hashes
-						numHashes -= power;
-
-						// one cache is filled with hashes
-						if (numCachesLeft != 0) numCachesLeft--;
-					}
-				}
-			}
-		}
-		else
+		size_t PUs = 0;
+		findChildrenByType(obj, HWLOC_OBJ_PU, [&PUs](hwloc_obj_t found) { PUs++; } );
+
+		//Strange case, but we will handle it silently, surely there must be one PU somewhere?
+		if(PUs == 0)
+			return;
+
+		if(obj->attr->cache.size == 0)
 		{
-			for (int i = 0; i < obj->arity; i++)
-				getConfig(topology, obj->children[i], numHashes, numCachesLeft);
+			//We will always have one child if PUs > 0
+			if(obj->children[0]->type != HWLOC_OBJ_CACHE)
+				throw(std::runtime_error("The CPU doesn't seem to have a cache."));
+
+			//Try our luck with lower level caches
+			for(size_t i=0; i < obj->arity; i++)
+				proccessTopLevelCache(obj->children[i]);
+			return;
 		}
-	}
 
-	inline void findCache(hwloc_topology_t topology, hwloc_obj_t obj)
-	{
-		if (obj->type == HWLOC_OBJ_CACHE)
+		size_t cacheSize = obj->attr->cache.size;
+		if(!isCacheExclusive(obj))
 		{
-			size_t cacheSize = obj->attr->cache.size;
-			size_t numHashL3 = ( cacheSize + m_scratchPadMemSize/ 2llu ) / m_scratchPadMemSize;
+			for(size_t i=0; i < obj->arity; i++)
+			{
+				hwloc_obj_t l2obj = obj->children[i];
+				//If L2 is exclusive and greater or equal to 2MB add room for one more hash
+				if(l2obj->type == HWLOC_OBJ_CACHE && l2obj->attr != nullptr && l2obj->attr->cache.size >= hashSize)
+					cacheSize += hashSize;
+			}
+		}
 
-			/* check cache is exclusive or inclusive */
-			const char* value = hwloc_obj_get_info_by_name(obj, "Inclusive");
+		std::vector<hwloc_obj_t> cores;
+		cores.reserve(16);
+		findChildrenByType(obj, HWLOC_OBJ_CORE, [&cores](hwloc_obj_t found) { cores.emplace_back(found); } );
 
+		size_t cacheHashes = (cacheSize + hashSize - 1) / hashSize;
```

Comment:
We should not round up if we are a few bytes over the 2MiB, instead we should use the some mechanism as in the old version [link](https://github.com/fireice-uk/xmr-stak-cpu/blob/dev/autoAdjustHwloc.hpp#L158) and only rounding up if `>=1Mib` is available.